src/azul/chalice.py

```diff
@@ -65,6 +66,30 @@ def __init__(self, app_name, unit_test=False, spec=None):
         else:
             self._specs: Optional[MutableJSON] = None
         super().__init__(app_name, debug=config.debug > 0, configure_logs=False)
+        # The last handler is called first. We want logging to surround all
+        # other handler calls.
+        self.register_middleware(self.logging_middleware, 'http')
+        self.register_middleware(self.authentication_middleware, 'http')
+
+    def logging_middleware(self, event, get_response):
+        self._log_request()
+        response = get_response(event)
+        self._log_response(response)
+        return response
+
+    def authentication_middleware(self, event, get_response):
+        # We use the handler's private method to generate the response in case
+        # authentication fails
+        handler = RestAPIEventHandler(self.routes,
```

Comment:
> The reason for switching to middleware is to reduce the hackyness of the current approach

This isn't the main reason for switching, but rather a hopeful side-effect. The reason we switched was because the method we were previously overriding in `Chalice()` no longer existed. Middleware seemed like the simplest way to get the same functionality.

> This is no less hacky than what we already do, it's less efficient and more surprising (two handler objects).

I agree, instantiating the handler again is hacky. The alternatives are:
- to duplicate the body of `_get_view_function_response()` (maybe not a bad idea) or
- subclass `RestAPIEventHander()` and change the behavior. 
I'll explain why below.

> I think the global error handler (_global_error_handler) is the outermost handler so the translation of any exception raised in _authenticate would still happen, even if we invoked it explicitly.

This is incorrect. `_global_error_hander()` simply packages the exception in a 500 response via `_unhandled_exception_to_response()`.

Anyway, assuming all of this stands, would you consider resetting the review counter?

Reply:
I'll bring this up in PL tomorrow.

Reply:
> > The reason for switching to middleware is to reduce the hackyness of the current approach
> 
> This isn't the main reason for switching, but rather a hopeful side-effect. The reason we switched was because the method we were previously overriding in `Chalice()` no longer existed. Middleware seemed like the simplest way to get the same functionality.

Let me rephrase that. Authentication is a text book use case of middleware. We should be able to implement authentication using middleware without resorting to hacks.

> > I think the global error handler (_global_error_handler) is the outermost handler so the translation of any exception raised in _authenticate would still happen, even if we invoked it explicitly.
> 
> This is incorrect. `_global_error_hander()` simply packages the exception in a 500 response via `_unhandled_exception_to_response()`.

But the "packaging" of the exception in a 500 response is exactly the translation I was talking about. I don't see where my statement is incorrect. What are we not getting from the global handler that we're getting from your current approach?

> Anyway, assuming all of this stands, would you consider resetting the review counter?

I'll decide when it is clear that A) there is no better implementation or B) the implementation we end up using couldn't possibly have been implemented by you initially, because information wasn't available to you or I changed the specification mid-PR.